File: androidApp/src/main/kotlin/it/mensa/app/features/auth/PasskeyAuthenticator.kt
package it.mensa.app.features.auth

import android.content.Context
import android.os.Build
import androidx.credentials.CreatePublicKeyCredentialRequest
import androidx.credentials.CreatePublicKeyCredentialResponse
import androidx.credentials.CredentialManager
import androidx.credentials.GetCredentialRequest
import androidx.credentials.GetPublicKeyCredentialOption
import androidx.credentials.PublicKeyCredential
import androidx.credentials.exceptions.CreateCredentialCancellationException
import androidx.credentials.exceptions.CreateCredentialException
import androidx.credentials.exceptions.GetCredentialCancellationException
import androidx.credentials.exceptions.GetCredentialException
import androidx.credentials.exceptions.NoCredentialException

/**
 * Prompt di sistema per creare e usare le passkey, via Credential Manager.
 *
 * Il bridge qui e' sottile perche' `CreatePublicKeyCredentialRequest` e
 * `GetPublicKeyCredentialOption` prendono e restituiscono direttamente JSON
 * WebAuthn: le options arrivano dal backend gia' nella forma giusta e la
 * risposta si rigira al backend cosi' com'e'. (Su iOS invece ASAuthorization
 * vuole i campi decodificati, e il JSON di risposta va ricostruito a mano.)
 *
 * Vive nell'app module e non nello `shared` perche' serve un [Context] di
 * Activity per mostrare il prompt.
 */
class PasskeyAuthenticator(private val activityContext: Context) {

    private val credentialManager = CredentialManager.create(activityContext)

    /**
     * Esegue l'assertion. [requestOptionsJson] e' il JSON che arriva da
     * `/api/cs/auth-with-passkey/begin`.
     */
    suspend fun assert(requestOptionsJson: String): PasskeyPromptResult {
        if (!isSupported) return PasskeyPromptResult.Unsupported

        return try {
            val response = credentialManager.getCredential(
                context = activityContext,
                request = GetCredentialRequest(
                    listOf(GetPublicKeyCredentialOption(requestJson = requestOptionsJson)),
                ),
            )
            val credential = response.credential as? PublicKeyCredential
                ?: return PasskeyPromptResult.Failed("unexpected credential type")
            PasskeyPromptResult.Success(credential.authenticationResponseJson)
        } catch (_: NoCredentialException) {
            // Nessuna passkey su questo dispositivo. Non e' un errore: e' il caso
            // di chi ha la passkey su un altro telefono, o non l'ha mai creata.
            PasskeyPromptResult.NoCredential
        } catch (_: GetCredentialCancellationException) {
            PasskeyPromptResult.Cancelled
        } catch (e: GetCredentialException) {
            PasskeyPromptResult.Failed(e.errorMessage?.toString() ?: e.type)
        }
    }

    /**
     * Crea una passkey. [creationOptionsJson] arriva da `/api/cs/passkeys/begin`
     * e porta gia' `excludeCredentials` popolato da Zitadel: se su questo
     * dispositivo esiste gia' una passkey per l'account, e' l'authenticator a
     * rifiutare, e la cosa emerge qui come [PasskeyPromptResult.AlreadyRegistered].
     */
    suspend fun create(creationOptionsJson: String): PasskeyPromptResult {
        if (!isSupported) return PasskeyPromptResult.Unsupported

        return try {
            val response = credentialManager.createCredential(
                context = activityContext,
                request = CreatePublicKeyCredentialRequest(requestJson = creationOptionsJson),
            )
            val created = response as? CreatePublicKeyCredentialResponse
                ?: return PasskeyPromptResult.Failed("unexpected create response type")
            PasskeyPromptResult.Success(created.registrationResponseJson)
        } catch (_: CreateCredentialCancellationException) {
            PasskeyPromptResult.Cancelled
        } catch (e: CreateCredentialException) {
            if (e.isAlreadyRegistered()) {
                PasskeyPromptResult.AlreadyRegistered
            } else {
                PasskeyPromptResult.Failed(e.errorMessage?.toString() ?: e.type)
            }
        }
    }

    /**
     * Il duplicato si riconosce dal `InvalidStateError` del DOM, che e' quello
     * che WebAuthn usa quando `excludeCredentials` fa match.
     *
     * Il match e' su stringhe e non sulla gerarchia `DomError` di proposito: e'
     * la parte di API di Credential Manager che si e' mossa piu' fra le versioni,
     * mentre `type` ed `errorMessage` sono stabili. Stessa scelta fatta lato
     * server con i codici d'errore di Zitadel (vedi `isPasswordError` in
     * MensaCore/tools/zauth/login.go).
     */
    private fun CreateCredentialException.isAlreadyRegistered(): Boolean {
        val haystack = "$type ${errorMessage ?: ""}".lowercase()
        return "invalidstateerror" in haystack || "already registered" in haystack
    }

    companion object {
        /**
         * Credential Manager esiste da API 21 ma le passkey servono API 28+.
         * Sotto quella soglia il bottone non va nemmeno mostrato: `minSdk` qui
         * e' 24, quindi il caso e' reale.
         */
        val isSupported: Boolean
            get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.P

        /** Stesso limite applicato dal server (`passkeyNameMaxLen`). */
        private const val MAX_NAME_LEN = 100

        /**
         * Etichetta con cui registrare la passkey: il modello del dispositivo.
         * Senza, una lista con tre voci "Passkey" e' inutilizzabile.
         */
        fun deviceLabel(): String {
            val model = Build.MODEL.orEmpty().trim()
            val manufacturer = Build.MANUFACTURER.orEmpty().trim()
            val composed = when {
                model.isEmpty() && manufacturer.isEmpty() -> "Android"
                model.startsWith(manufacturer, ignoreCase = true) -> model
                manufacturer.isEmpty() -> model
                else -> "$manufacturer $model"
            }
            return composed.take(MAX_NAME_LEN)
        }
    }
}

/** Esito di un prompt di sistema. */
sealed interface PasskeyPromptResult {
    /** [credentialJson] va rigirato al backend cosi' com'e'. */
    data class Success(val credentialJson: String) : PasskeyPromptResult

    /** L'utente ha chiuso il prompt. Nessun messaggio: non e' un errore. */
    data object Cancelled : PasskeyPromptResult

    /** Nessuna passkey utilizzabile su questo dispositivo. */
    data object NoCredential : PasskeyPromptResult

    /** Su questo dispositivo esiste gia' una passkey per l'account. */
    data object AlreadyRegistered : PasskeyPromptResult

    /** Dispositivo troppo vecchio (API < 28). */
    data object Unsupported : PasskeyPromptResult

    data class Failed(val message: String?) : PasskeyPromptResult
}

File: iosApp/iosApp/Features/Login/PasskeyAuthenticator.swift
import AuthenticationServices
import Foundation
import UIKit

/// Ponte fra le WebAuthn options che arrivano dal backend e `ASAuthorization`.
///
/// Su Android l'equivalente e' banale, perche' Credential Manager accetta e
/// restituisce direttamente il JSON WebAuthn. `ASAuthorization` no: vuole i
/// campi decodificati in ingresso e restituisce oggetti tipizzati in uscita,
/// quindi il JSON di risposta va ricostruito a mano. Tutto quel lavoro sta qui,
/// cosi' i ViewModel vedono solo `String` in e `String` out.
///
/// L'encoding e' base64url senza padding in entrambe le direzioni: e' il formato
/// che WebAuthn richiede ed e' anche quello che Zitadel emette e si aspetta
/// (vedi il commento in MensaCore/tools/zauth/webauthn_json.go).
@MainActor
final class PasskeyAuthenticator: NSObject {

    /// Deve combaciare con `zauth.PasskeyRPID` lato server e con il dominio che
    /// serve l'apple-app-site-association. Resta come fallback: l'`rpId` vero
    /// arriva dentro le options.
    static let defaultRelyingPartyID = "auth.mensa.it"

    enum Failure: Error {
        /// Le options non sono decodificabili: bug di protocollo, non un caso utente.
        case malformedOptions(String)
        /// L'utente ha annullato, **oppure** non c'e' nessuna passkey su questo
        /// device. iOS riporta i due casi con lo stesso `ASAuthorizationError.canceled`
        /// e non sono distinguibili: chi chiama deve trattarli allo stesso modo,
        /// cioe' riportando l'utente alla password senza mostrare un errore.
        case cancelledOrNoCredential
        case authorizationFailed(Error)
        /// Il registration non ha restituito l'attestation object.
        case missingAttestation
    }

    private var continuation: CheckedContinuation<ASAuthorizationCredential, Error>?

    // MARK: - Login

    /// Esegue l'assertion e ritorna il `PublicKeyCredential` JSON da mandare a
    /// `/api/cs/auth-with-passkey/finish`.
    func assert(requestOptionsJSON: String) async throws -> String {
        let options = try AssertionOptions(json: requestOptionsJSON)

        let provider = ASAuthorizationPlatformPublicKeyCredentialProvider(
            relyingPartyIdentifier: options.rpID ?? Self.defaultRelyingPartyID
        )
        let request = provider.createCredentialAssertionRequest(challenge: options.challenge)

        if !options.allowedCredentialIDs.isEmpty {
            request.allowedCredentials = options.allowedCredentialIDs.map {
                ASAuthorizationPlatformPublicKeyCredentialDescriptor(credentialID: $0)
            }
        }
        if let preference = options.userVerification {
            request.userVerificationPreference = .init(rawValue: preference)
        }

        // preferImmediatelyAvailableCredentials: senza una passkey locale
        // fallisce subito invece di aprire uno sheet che qui non porterebbe da
        // nessuna parte — abbiamo scelto di non supportare il flusso QR
        // cross-device, quindi l'unica alternativa e' la password.
        let credential = try await perform(
            requests: [request],
            options: .preferImmediatelyAvailableCredentials
        )

        guard let assertion = credential as? ASAuthorizationPlatformPublicKeyCredentialAssertion else {
            throw Failure.authorizationFailed(Failure.malformedOptions("unexpected credential type"))
        }

        return try encodeJSON([
            "id": assertion.credentialID.base64URLString,
            "rawId": assertion.credentialID.base64URLString,
            "type": "public-key",
            "authenticatorAttachment": "platform",
            "clientExtensionResults": [String: String](),
            "response": [
                "clientDataJSON": assertion.rawClientDataJSON.base64URLString,
                "authenticatorData": assertion.rawAuthenticatorData.base64URLString,
                "signature": assertion.signature.base64URLString,
                "userHandle": assertion.userID.base64URLString,
            ],
        ])
    }

    // MARK: - Registrazione

    /// Crea la passkey e ritorna il `PublicKeyCredential` JSON da mandare a
    /// `/api/cs/passkeys/finish`.
    func create(creationOptionsJSON: String) async throws -> String {
        let options = try CreationOptions(json: creationOptionsJSON)

        let provider = ASAuthorizationPlatformPublicKeyCredentialProvider(
            relyingPartyIdentifier: options.rpID ?? Self.defaultRelyingPartyID
        )
        let request = provider.createCredentialRegistrationRequest(
            challenge: options.challenge,
            name: options.userName,
            userID: options.userID
        )
        if let preference = options.userVerification {
            request.userVerificationPreference = .init(rawValue: preference)
        }

        // excludedCredentials esiste solo da iOS 17.4. Sotto quella soglia
        // l'utente puo' creare una seconda passkey per lo stesso account sullo
        // stesso device: Zitadel le accetta entrambe, e la schermata di gestione
        // permette di rimuovere il duplicato. Nessun dato a rischio, solo una
        // lista piu' disordinata.
        if #available(iOS 17.4, *), !options.excludedCredentialIDs.isEmpty {
            request.excludedCredentials = options.excludedCredentialIDs.map {
                ASAuthorizationPlatformPublicKeyCredentialDescriptor(credentialID: $0)
            }
        }

        let credential = try await perform(requests: [request], options: [])

        guard let registration = credential as? ASAuthorizationPlatformPublicKeyCredentialRegistration else {
            throw Failure.authorizationFailed(Failure.malformedOptions("unexpected credential type"))
        }
        guard let attestation = registration.rawAttestationObject else {
            throw Failure.missingAttestation
        }

        return try encodeJSON([
            "id": registration.credentialID.base64URLString,
            "rawId": registration.credentialID.base64URLString,
            "type": "public-key",
            "authenticatorAttachment": "platform",
            "clientExtensionResults": [String: String](),
            "response": [
                "clientDataJSON": registration.rawClientDataJSON.base64URLString,
                "attestationObject": attestation.base64URLString,
            ],
        ])
    }

    // MARK: - Esecuzione

    private func perform(
        requests: [ASAuthorizationRequest],
        options: ASAuthorizationController.RequestOptions
    ) async throws -> ASAuthorizationCredential {
        try await withCheckedThrowingContinuation { continuation in
            self.continuation = continuation
            let controller = ASAuthorizationController(authorizationRequests: requests)
            controller.delegate = self
            controller.presentationContextProvider = self
            controller.performRequests(options: options)
        }
    }

    private func resume(with result: Result<ASAuthorizationCredential, Error>) {
        guard let continuation else { return }
        self.continuation = nil
        continuation.resume(with: result)
    }

    private func encodeJSON(_ payload: [String: Any]) throws -> String {
        let data = try JSONSerialization.data(withJSONObject: payload, options: [])
        guard let json = String(data: data, encoding: .utf8) else {
            throw Failure.malformedOptions("credential payload is not valid UTF-8")
        }
        return json
    }
}

// MARK: - Delegate

extension PasskeyAuthenticator: ASAuthorizationControllerDelegate {
    func authorizationController(
        controller: ASAuthorizationController,
        didCompleteWithAuthorization authorization: ASAuthorization
    ) {
        resume(with: .success(authorization.credential))
    }

    func authorizationController(
        controller: ASAuthorizationController,
        didCompleteWithError error: Error
    ) {
        if let authError = error as? ASAuthorizationError, authError.code == .canceled {
            resume(with: .failure(Failure.cancelledOrNoCredential))
            return
        }
        resume(with: .failure(Failure.authorizationFailed(error)))
    }
}

extension PasskeyAuthenticator: ASAuthorizationControllerPresentationContextProviding {
    func presentationAnchor(for controller: ASAuthorizationController) -> ASPresentationAnchor {
        // Stessa risoluzione della scena usata da StripeService.
        let scene = UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .first { $0.activationState == .foregroundActive }
            ?? UIApplication.shared.connectedScenes
                .compactMap { $0 as? UIWindowScene }
                .first
        return scene?.keyWindow ?? scene?.windows.first ?? ASPresentationAnchor()
    }
}

// MARK: - Decodifica delle options

/// Sottoinsieme di `PublicKeyCredentialRequestOptions` che serve ad ASAuthorization.
private struct AssertionOptions {
    let challenge: Data
    let rpID: String?
    let allowedCredentialIDs: [Data]
    let userVerification: String?

    init(json: String) throws {
        let root = try PasskeyOptionsDecoder.object(from: json)

        guard let challenge = PasskeyOptionsDecoder.data(root["challenge"]) else {
            throw PasskeyAuthenticator.Failure.malformedOptions("challenge")
        }
        self.challenge = challenge
        self.rpID = root["rpId"] as? String
        self.userVerification = root["userVerification"] as? String
        self.allowedCredentialIDs = PasskeyOptionsDecoder.credentialIDs(root["allowCredentials"])
    }
}

/// Sottoinsieme di `PublicKeyCredentialCreationOptions`.
private struct CreationOptions {
    let challenge: Data
    let rpID: String?
    let userName: String
    let userID: Data
    let excludedCredentialIDs: [Data]
    let userVerification: String?

    init(json: String) throws {
        let root = try PasskeyOptionsDecoder.object(from: json)

        guard let challenge = PasskeyOptionsDecoder.data(root["challenge"]) else {
            throw PasskeyAuthenticator.Failure.malformedOptions("challenge")
        }
        self.challenge = challenge
        self.rpID = (root["rp"] as? [String: Any])?["id"] as? String

        guard let user = root["user"] as? [String: Any] else {
            throw PasskeyAuthenticator.Failure.malformedOptions("user")
        }
        guard let userID = PasskeyOptionsDecoder.data(user["id"]) else {
            throw PasskeyAuthenticator.Failure.malformedOptions("user.id")
        }
        self.userID = userID
        // `name` e' il login name mostrato nel prompt di sistema e salvato nel
        // portachiavi; se manca, meglio una stringa vuota che far fallire tutto.
        self.userName = (user["name"] as? String) ?? ""

        self.userVerification =
            (root["authenticatorSelection"] as? [String: Any])?["userVerification"] as? String
        self.excludedCredentialIDs = PasskeyOptionsDecoder.credentialIDs(root["excludeCredentials"])
    }
}

private enum PasskeyOptionsDecoder {
    static func object(from json: String) throws -> [String: Any] {
        guard let data = json.data(using: .utf8),
              let root = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            throw PasskeyAuthenticator.Failure.malformedOptions("root object")
        }
        return root
    }

    static func data(_ value: Any?) -> Data? {
        guard let encoded = value as? String else { return nil }
        return Data(base64URLEncoded: encoded)
    }

    static func credentialIDs(_ value: Any?) -> [Data] {
        guard let list = value as? [[String: Any]] else { return [] }
        return list.compactMap { data($0["id"]) }
    }
}

// MARK: - base64url

private extension Data {
    /// base64url senza padding, come richiesto da WebAuthn JSON.
    var base64URLString: String {
        base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    /// Tollera anche il padding e l'alfabeto standard, cosi' un cambio di
    /// encoding lato server non ci rompe silenziosamente.
    init?(base64URLEncoded value: String) {
        var normalized = value
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        while normalized.count % 4 != 0 { normalized += "=" }
        guard let decoded = Data(base64Encoded: normalized) else { return nil }
        self = decoded
    }
}
